### ISSUE TITLE
Padding in convolve2NN

### ISSUE THREAD

**Issue by geq1t** [arrayfire/arrayfire#2817]
Hi,

Is there any particular reason why I can't use padding of zero in convolve2NN() to get a valid convolution? Right now, the padding argument must not contain any non-zero elements, which feels overly restrictive to me.

Thanks for any insights you can give me.

Regards,
Gé Vissers

**Comment by 9prady9** [comment]
@geq1t Do you mean `convolve2NN(signal, filter, strides, dim4(0, 0), dilation)` is not working ?

I couldn't find any checks in at the entry point of API that ensures non-zero values in padding argument.

 Can you please show me the `convolve2NN` with necessary parameter values of `dim4` arguments.

**Comment by geq1t** [comment]
> @geq1t Do you mean `convolve2NN(signal, filter, strides, dim4(0, 0), dilation)` is not working ?

That is eactly what I mean:
```
#include <arrayfire.h>

int main()
{
    af::array signal = af::randu(5, 5);
    af::array filter = af::randu(3, 3);
    af::dim4 strides(1, 1), dilation(1, 1);
    af::dim4 padding(0, 0, 1, 1);

    convolve2NN(signal, filter, strides, padding, dilation);
}
```
gives the following output
```
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Invalid input argument:202):
In function af_err af_convolve2_nn(void**, af_array, af_array, unsigned int, const dim_t*, unsigned int, const dim_t*, unsigned int, const dim_t*)
In file src/api/c/convolve.cpp:334
Invalid argument at index 5
Expected: padding_dims > 0 && padding_dims <= 2
 0# 0x00007F38D1A5FAFD in /usr/lib/libafopencl.so.3
 1# af_convolve2_nn in /usr/lib/libaf.so.3
 2# af::convolve2NN(af::array const&, af::array const&, af::dim4, af::dim4, af::dim4) in /usr/lib/libaf.so.3
 3# 0x000055E015B752CD in ./conv
 4# __libc_start_main in /usr/lib/libc.so.6
 5# 0x000055E015B750CE in ./conv

In function af::array af::convolve2NN(const af::array&, const af::array&, af::dim4, af::dim4, af::dim4)
In file src/api/cpp/convolve.cpp:58
Aborted (core dumped)
```

> I couldn't find any checks in at the entry point of API that ensures non-zero values in padding argument.
> 
> Can you please show me the `convolve2NN` with necessary parameter values of `dim4` arguments.

The check is in src/api/c/convolve.cpp, line 334:
```
        ARG_ASSERT(5, padding_dims > 0 && padding_dims <= 2);
```

**Comment by 9prady9** [comment]
Oh yea, I get what is going now. `dim4::ndims()` returns zero if the number of elements are zero. Let me look into what can be done here. Thanks for reporting it.

Note: You should be able to overcome this particular issue using C API though, but it would make your code more verbose and you would end up managing af_array just for that case. Not ideal or desirable I would think.

**Comment by geq1t** [comment]
> Oh yea, I get what is going now. `dim4::ndims()` returns zero if the number of elements are zero. Let me look into what can be done here. Thanks for reporting it.

Thanks you looking into this so quickly!

> Note: You should be able to overcome this particular issue using C API though, but it would make your code more verbose and you would end up managing af_array just for that case. Not ideal or desirable I would think.

Thanks, that could work. In my project I am using arrayfire from Rust through the arrayfire-rust bindings, though. If at all possible, I'd like to stay away from writing my own custom wrappers and the associated headaches :)

**Comment by geq1t** [comment]
(Closed the issue by mistake, sorry)

**Comment by 9prady9** [comment]
@geq1t If you are using rust-wrapper, may be we can handle it over there because rust wrapper is written on top of arrayfire's C-API so it could be handled over there as well. But I believe we need to verify if there is anything else that can be done in C++ API to avoid similar issue.

**Comment by 9prady9** [comment]
Fixed in #2820
